### PR TITLE
ddl: reuse chunk for copr-read and check context done

### LINKED ISSUE #39308: data inconsistency after adding index with ingest method
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、set @@global.tidb_ddl_enable_fast_reorg=1;
2、run tpcc or sysbench read_write
3、add index and after 1min，restart tidb1
4、stop workload
select count(1) from t use index (idx);
select count(1) from t ignore index (idx);
5、the count is inconsist

### 2. What did you expect to see? (Required)
use index and ignore index is inconsistent for table sbtest12, use index: 10007147, ignore index: 10003873
the count is the same of step 4
image

### 3. What did you see instead (Required)
the count is not the same of step 4
http://perf.pingcap.net:3000/d/WOw1JWmVk/ha-test-case-detail?orgId=1&refresh=10s&var-case_id=750084

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v6.5.0-alpha
Edition: Community
Git Commit Hash: 028c5eb8359d1f756b99e40f33c8f3f998f7769e
Git Branch: heads/refs/tags/v6.5.0-alpha
UTC Build Time: 2022-11-21 11:13:49
GoVersion: go1.19.2
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore

### LINKED ISSUE #39468: Ingest add index job cannot be cancelled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Create 100M table using sysbench.
2. Add index for the column c(char(120)).
3. Run `admin cancel ddl jobs`.
4. TiDB DDL hangs.

### 2. What did you expect to see? (Required)

The adding index job should be able to cancel.

### 3. What did you see instead (Required)

TiDB keeps report "[ddl] run the cancelling DDL job" constantly.

### 4. What is your TiDB version? (Required)

d2ace99d68

### LINKED ISSUE #39470: Ingest add index job consumes a lot of memory
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Create 100M table using sysbench.
2. Add index for the column c(char(120)).
3. The memory consumption keep increasing.

### 2. What did you expect to see? (Required)

It should be less than 8GB.

### 3. What did you see instead (Required)

The memory consumption increases to more than 15GB.

### 4. What is your TiDB version? (Required)

d2ace99d68

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #39468, close #39470

Problem Summary:

For #39468, when `ADMIN CANCEL DDL JOBS` is executed, the context is cancelled. However, the cop-request senders does not aware of this and keep sending to the results channel. As a result, `WaitGroup.Wait()` blocks forever because the cop-request senders never exit.

For #39470, I found that the chunk introduced in #39382 is never reused and the runtime GC worker cannot collect it because `sync.Pool` keeps the reference. This PR uses a different way to fix #39308: it clones everything for each datum to prevent sharing the same memory.

### What is changed and how it works?

See the problem summary.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Before this PR:

image

After this PR:

image

Before this PR:

```
+--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE               | SCHEMA_STATE         | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE         |
+--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
|     85 | sbtest  | sbtest1    | add index /* ingest */ | write reorganization |        76 |       78 |  39063370 | 2022-11-29 21:19:31 | 2022-11-29 21:19:31 | NULL                | cancelling    |
+--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
```

After this PR:
```
+--------+---------+------------+------------------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE               | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE         |
+--------+---------+------------+------------------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
|     88 | sbtest  | sbtest1    | add index /* ingest */ | none         |        76 |       78 |         0 | 2022-11-29 21:55:55 | 2022-11-29 21:55:55 | 2022-11-29 21:56:34 | rollback done |
+--------+---------+------------+------------------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+

ERROR 8214 (HY000) at line 2: Cancelled DDL job
```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/index_cop.go [hunk 1]
```diff
@@ -155,19 +166,20 @@ func newCopReqSenderPool(ctx context.Context, copCtx *copContext, startTS uint64
...
-		binding: generic.NewSyncMap[*[]*indexRecord, *chunk.Chunk](4),
 	}
 }
 
 func (c *copReqSenderPool) sendTask(task *reorgBackfillTask) {
 	c.tasksCh <- task
 }
 
+func (c *copReqSenderPool) sendResult(ctx context.Context, result idxRecResult) {
+	select {
```
Comment (Benjamin2037) [inline ddl/index_cop.go:177]: Here is the code for `Admin Cancel`?

File: ddl/index_cop.go [hunk 2]
```diff
@@ -122,24 +121,36 @@ func (c *copReqSender) run() {
 			zap.Int("id", task.id), zap.String("task", task.String()))
 		rs, err := p.copCtx.buildTableScan(p.ctx, p.startTS, task.startKey, task.excludedEndKey())
 		if err != nil {
-			p.resultsCh <- idxRecResult{id: task.id, err: err}
+			p.sendResult(c.ctx, idxRecResult{id: task.id, err: err})
 			return
 		}
 		var done bool
 		var total int
 		for !done {
```
Comment (Benjamin2037) [inline ddl/index_cop.go:129]: when `admin cancel` issued, it still need wait fetchTableScanResult done?

Comment (Benjamin2037) [reply ddl/index_cop.go:129]: ok，I see, this loop is by task.

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]
it work. But we maybe have a better method to solve. We have possiblity to use pool to resue object. It can be removed from ```binding``` and set ```Finalizer``` to put into ```pool``` after use. We can try it after releasing v6.5.

**Review by wjhuang2016 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Comment by hawkingrei** [comment]
> /run-check_dev_2

I have try to revert #39503 by #39521. But this task still fail.

File: ddl/index_cop.go [hunk 3]
```diff
@@ -401,7 +413,7 @@ func (c *copContext) fetchTableScanResult(ctx context.Context, result distsql.Se
 		return nil, false, errors.Trace(err)
 	}
 	if chk.NumRows() == 0 {
-		return buf, true, nil
```
Comment (Benjamin2037) [inline ddl/index_cop.go:404]: Here means finished scan, we closed result to release mem usage? how about   before err return should we also have safeguard to release mem?

**Pull request merged**